pkg/service/connections.go

```
@@ -125,9 +125,33 @@ func (s *Service) TerminateUDPConnectionsToBackend(l3n4Addr *lb.L3n4Addr) error
 
 // backendConnectionHandler is added for dependency injection in tests.
 type backendConnectionHandler struct {
-	logger *slog.Logger
+	sockets.SocketDestroyer
+	logger     *slog.Logger
+	useNetlink bool
+}
+
+func newBackendConnectionHandler(l *slog.Logger, sockTermFilter *filter.SockTermFilterMap) *backendConnectionHandler {
+	return &backendConnectionHandler{
+		SocketDestroyer: &sockets.BPFSocketDestroyer{
+			SockTermFilter: sockTermFilter,
+			Logger:         l,
+		},
+		logger: l,
+	}
 }
 
 func (h backendConnectionHandler) Destroy(filter sockets.SocketFilter) error {
-	return sockets.Destroy(logging.DefaultSlogLogger, filter)
+	err := h.SocketDestroyer.Destroy(filter)
+	if err != nil && !h.useNetlink {
+		h.logger.Error("error while attempting to use the BPF socket destroyer. Is bpf_sock_destroy supported on the current kernel? "+
```

This will happen each time a destroy call is issued, meaning we'll continue to emit these errors.  I'd say its likely better to probe for this functionality once at startup and then use that.

As well, this should be info level as it's it just means that the kernel doesn't yet support the new method (also errors will likely break CI error log checks in e2e-* tests).

---

As an example, we probe for bpf_map_lookup_batch support in the kernel once [here](https://github.com/cilium/cilium/blob/main/pkg/maps/ctmap/ctmap.go#L408-L410), then always use the supported method in the future.

---

Replied to this below. I was trying out GitHub's "reply by email" feature. Looks like it doesn't do threading -_-, sorry about that.

---

>As well, this should be info level as it's it just means that the kernel doesn't yet support the new method (also errors will likely break CI error log checks in e2e-* tests).

Yup, agree with Tom's comment.

---

>  I have some vague doubts that unexpected issues could arise around the lack of CO-RE for struct fields that I highlighted in the issue description.

> I tried using Co-Re helpers around field accesses but had limited success. While I could use them to get a pointer to sk, I couldn't pass that pointer to the bpf_sock_destroy() kfunc without the verifier complaining about unsafe pointers or some such thing. This could use some further investigation, but at least right now I'm not sure how to get this to work. So, I guess there's some risk of the field accesses breaking on newer kernels if offsets change in struct bpf_iter__udp. This probably warrants some discussion.

I see now that if an attempt fails we permanently fall back to using the netlink based one.  

Yeah, as mentioned we do  test against a lot of different kernel versions and configurations, including bpf-next, so such a regressions *should likely* be caught.   However, I think we only have this kernel coverage for the e2e tests (cc: @aditighag do we have any specific coverage of the svcLB socket termination in e2e-tests).

To this, because we'll need to make this Info level to avoid breaking CI, a worry with being overzealous with falling back is that in such a case that there is a regression it may go unnoticed for a while since we'd quietly begin fall back to the netlink based one.

---

It's also good to keep in mind that this change would be landing in the v1.19 release (we're currently just starting the v1.18 release candidates), so there would be a good amount of time for these changes to settle.

---

> To this, because we'll need to make this Info level to avoid breaking CI, a worry with being overzealous with falling back is that in such a case that there is a regression it may go unnoticed for a while since we'd quietly begin fall back to the netlink based one.

Yeah, fair point; i'm convinced. I'll rework things a little bit to probe up front for the kfunc and instantiate the correct destroyer accordingly.

---

@tommyp1ckles The latest push reworks this a bit. On startup, a new method of `BPFSocketDestroyer` called `IsSupported()` tests to see if the `bpf_sock_destroy()` kfunc is available. 

> To this, because we'll need to make this Info level to avoid breaking CI, a worry with being overzealous with falling back is that in such a case that there is a regression it may go unnoticed for a while since we'd quietly begin fall back to the netlink based one.

Instead of silently falling back to the netlink socket destroyer if an error occurs, it will fail to start up. This new approach should prevent failures from going unnoticed for a long time.

@joamaki made [some changes](https://github.com/cilium/cilium/pull/39012/commits/26593a1edc7b21c1becc02c84b5f694908ec4cc2) recently that will need to be merged to make the UDP socket destroyer stuff work again, but in the meantime feel free to look at the general approach and let me know what you think.